### PR TITLE
fix(helm): allow a previously failed release to be upgraded

### PR DESCRIPTION
**Pull request by yinzara**
This PR changes the behavior of "helm upgrade" command when there have been no successful and at least one failed release.  This allows the user to recover from a failed install (or a partially failed install) without uninstalling the release.   The user can now decide to decide to ignore the "no deployed releases" errors and allow it to upgrade a failed release.

There were no test cases that tested the failure scenarios of upgrading a failed/missing release so I added them and also added additional test cases that test the new functionality.

Fixes issue #5595

Signed-off-by: Matt Morrissete <yinzara@gmail.com>

### CODE REVIEW

**Comment by bacongobbler** [comment]
Taking a quick glance at the feature flag description...

> This is unsafe in production

If it's unsafe for production use, why are we advocating for others to use this?

I ask this because when we discuss introducing new features to Helm, we aim to implement features that can be used safely in production. If there are features unsafe for production use, they are temporarily hidden behind a feature gate until the feature is considered stable.

Why should we ship this?

**Comment by yinzara** [comment]
I only added the warning because it also exists on the original replace flag.  I honestly don't believe it's "unsafe for production".   I would more say it's smarter to only use this flag when necessary after you have determined ignoring the "no deployed releases" error is safe (i.e. you intended to upgrade a failed release).   I'm happy to reword.  Right now there is no way to recover from a failed release other than removing that release entirely (which may not be what you want).

I will use "--replace" in my development automated build process but will use "--atomic" in production and let me explain why.

It's very common in our automated build system for a new microservice to fail its first deployment (because of a mistake by the dev or otherwise).  We would like this solved by pushing another commit that fixes the deployment issue. Right now, without the "--replace" flag, we have to manually clean up every release if they failed on the first deploy (the main cause of pain outlined in #5595 ).  I would like it for our dev builds, that if a release fails, it remains in that state so I can debug what occurred with it but if I push a new commit that fixes the failure, that the deploy succeeds.

However in production, if a release fails, I would like that release to rollback so that my production environment remains pristine.  I'd like to know the release failed but I want my systems to remain in their prior-to-rollout state if that failure occurs.

Without this flag, there is no way to guarantee that.

**Comment by yinzara** [comment]
@bacongobbler Any chance we could get some love here (and on #7668 if you're cool with this PR)

**Comment by bacongobbler** [comment]
I’m busy with internal work this week. Won’t be able to review anything for a little while. Sorry!

**Comment by yinzara** [comment]
I friend called out that this PR seems similar to #7431 

I do not believe that #7431 solves the issue that "--replace" does and I believe "--replace" follows the spirit of the same flag in the "install" command without introducing any new concepts and without violating the general process flow of the current helm upgrade nor introducing a lot of code changes.

**Comment by dudicoco** [comment]
@yinzara does this flag only replaces the failed release or also the resources within the cluster?

**Comment by yinzara** [comment]
This flag acts as if the most recent failed release (if there are no successful ones) was successful.  i.e. any resources that it should be deployed will be synchronized with the rendered chart.  So let's say Chart V1 had resource A, B, C and Chart V2 had resource C, D and E.

You install Chart V1 by using "helm upgrade --install --replace release-name chart-name ..." and it fails however A and C were installed correctly and B failed to install.

You then update the source to V2 using the same command which is missing resources A and B but has new resources D and E.  Helm will act as if the first release was successful and assume A, B and C were all installed.   

It will attempt to delete resources A and C, upgrade B to match V2's config, and install D and E with V2s config.  Resource A will be deleted as it was successfully created in V1.  C will attempt delete but will fail and print a warning to the log that helm attempted to delete a non-existent resource. Helm will patch B and install D and E with the config from V2. 

Maintaining existing functionality, if resource E was already found in the cluster when running the upgrade command on V2, helm will error indicating there was a resource not managed by the chart that already existed in the cluster.

**Comment by dudicoco** [comment]
@yinzara it's not very straight forward to understand this behaviour. How would you describe in in the docs?

**Comment by yinzara** [comment]
I'm confused why that behavior is not straightforward. It's literally how Helm always works in a "helm upgrade" scenario even without the "--replace" flag.  The only thing the replace flag does is in the case where you have no successful releases, it allows the most recent failed release to act as if it was successful.  See the actual changed code in my PR, all it does is ignore an error.  It does not actually change any functionality of how helm manages the upgrade. 

I also described above how I would use this in an automated build/release workflow and why I would use it over the "--atomic" flag.  What else would make this easier to swallow?

The documentation provided with the flag (as listed above is):
if no deployed version of the release is available, replace an uninstalled, pending install, or failed release which remains in the history. If no prior failed, uninstalled, pending install or deployed release is available, --replace will not install a new release unless --install is also specified

**Review by BrunoChauvet — APPROVED** [review]

**Comment by yinzara** [comment]
Any thing else needed to get this merged?

**Comment by yinzara** [comment]
@BrunoChauvet How do we get this merged?

**Comment by bacongobbler** [comment]
Thank you for the ping.

At this time, I'm focused on other work. I do not have enough time in my work schedule to review your code in the near future. That being said, there are plenty of other Helm maintainers who may have the time to review your work. I'd reach out and see if someone else is available to review this.

Sorry for the inconvenience.

**Comment by yinzara** [comment]
Thanks man! I appreciate it anyway.

How about @mattfarina ?

**Review by yashbhutwala — APPROVED** [review]

File: pkg/action/upgrade.go
Comment (lukaspj) [inline pkg/action/upgrade.go:130]: Would be nice if this were more statically typed. Can `errors.Is` be used?

Comment (yinzara) [reply pkg/action/upgrade.go:130]: The error is not a constant.  It's a string built with the name of the release in it (hence why I used strings.Contains.  Is there another method I don't know?

**Comment by yinzara** [comment]
Who can merge this pull request? There are multiple approvals.

**Comment by yinzara** [comment]
@technosophos Can we get this merged?

File: pkg/action/upgrade.go
Comment (lukaspj) [reply pkg/action/upgrade.go:130]: My apologies, was not familiar with the codebase, it just looked incredibly.. Wrong..
But alas this seems to be the intended design throughout the codebase, so nothing wrong with it here.

**Comment by technosophos** [comment]
Can you describe the cases in which this is preferable to `--atomic`?

**Comment by lukaspj** [comment]
We had an edge-case where the new Helm installation, run by CD, using --atomic failed and it wanted to rollback, however the version it rolled back to failed as well.

Helm retains a list of deployments with max size 10 it seems and uses it to rollback. The atomic action of install, rollback added 2 failed deployments to that list.

At some point we messed up so badly that we failed 5 upgrades in a row, meaning 5*2 (due to rollbacks) failed deployments. So there were only failed deployments in the list.

The result was that upgrade, even with atomic, no longer worked because there were no successful deployment in the history so Helm would early out.

As I understand it, this change would alleviate such errors by doing the upgrade in spite of the lack of a successful deployment.

**Comment by yinzara** [comment]
@technosophos 
I'm reposting an earlier comment you may have missed that explains why "--atomic" does not solve this issue.

I will use "--replace" in my development automated build process but will use "--atomic" in production and let me explain why.

It's very common in our automated build system for a new microservice to fail its first deployment (because of a mistake by the dev or otherwise). We would like this solved by pushing another commit that fixes the deployment issue. Right now, without the "--replace" flag, we have to manually clean up every release if they failed on the first deploy (the main cause of pain outlined in #5595 ). I would like it for our dev builds, that if a release fails, it remains in that state so I can debug what occurred with it but if I push a new commit that fixes the failure, that the deploy succeeds.

However in production, if a release fails, I would like that release to rollback so that my production environment remains pristine. I'd like to know the release failed but I want my systems to remain in their prior-to-rollout state if that failure occurs.

The '--replace' flag is necessary in the above use case as there is no other combination of options that solves this issue.


Additionally, there is the case where a user has installed a large public helm chart (without --atomic) and is now in a state where the release only failed because of a timeout and not because of an actual error and without "helm upgrade --replace" there is no way to complete the installation of the chart without fully uninstalling the entire installation which most of the time is unnecessary.

**Comment by yinzara** [comment]
Alrighty, so I have removed the "--replace" option and made the behavior the default when upgrading a previously failed release with no prior successful releases.  I changed the errors to use typed errors though I went a little farther and added a StorageDriverError class to allow me to keep the error messages the same as they were before.

File: pkg/action/upgrade.go
Comment (yinzara) [inline pkg/action/upgrade.go:253]: I'm not sure why this is appearing.  I rebased against master.

**Comment by yinzara** [comment]
So sorry for all the force pushes. I made a mistake rebasing and it took me a while to get it fixed.

**Comment by technosophos** [comment]
Okay, we are reviewing this new version. I closed my similar PR in favor of this one. The last major thing that we (the core maintainers) are trying to decide upon is whether we want to merge this for 3.2. I recognize that by asking for this to be the default behavior, we're taking on a little bit higher of a risk. But I am hoping that if we get this into 3.2 in time for it's beta, we will get plenty of testing time before the final 3.2.0 comes out.

**Review by technosophos — APPROVED** [review]
I am good with this revision. I believe @adamreese is reviewing it as well.

**Comment by jawlitkp** [comment]
I am testing 3.2 right now. so I can see it always keeps one good previous revision if any which is awesome. but the other scenario where you don't have any good previous revision, it still fails to upgrade. So is that expected?

**Comment by bacongobbler** [comment]
this PR has not yet been merged. Yes, that is the current behaviour this PR is trying to fix.

**Comment by jawlitkp** [comment]
Ohh ok Thanks! That's the fix I am looking for. Any ETA?

**Comment by technosophos** [comment]
Just waiting for a second LGTM. I co-reviewed with Adam, so I will remind him to sign off on this one.

**Review by adamreese — APPROVED** [review]

**Comment by yinzara** [comment]
Can't tell you how excited I am to see both my PRs accepted :) Can't wait to upgrade my workflows with 3.2!

**Comment by technosophos** [comment]
I have put this on 3.2.0, so this will make it into release 3.2.0-RC.2.

**Comment by jawlitkp** [comment]
wow!! Cant wait. Thanks Guys !!

**Comment by hiro5id** [comment]
Our CI is **literally** stuck right now because of this error.  Waiting on pins and needles when this is released. Hoping that by Monday we'll be able to use this feature, even if it is in a RC.

**Comment by nareshnayini** [comment]
Hey, general question. Would this be updated on helm install [page](https://helm.sh/docs/intro/install/) when it's ready though it's in RC.

**Comment by bacongobbler** [comment]
Feel free to test this PR with the canary release: https://helm.sh/docs/intro/install/#from-canary-builds

**Comment by nareshnayini** [comment]
gotcha, thank you.

**Comment by nareshnayini** [comment]
Tested this, seems the issue fixed 🎉

**Comment by hiro5id** [comment]
This didn't work for us, we used the latest canary version from here:
`https://get.helm.sh/helm-canary-linux-amd64.tar.gz`
and ran our usual command to upgrade:
`helm upgrade cce ./chart-path/ --set image.tag=xxx -i -n app-namespace --atomic --cleanup-on-fail`
But we still get error:
`Error: UPGRADE FAILED: "cce" has no deployed releases`

when we do `helm list` we see this:
```
NAME    	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART           	APP VERSION
cce     	cce      	981     	2020-04-16 14:15:42.732894084 +0000 UTC	failed  	cce-k8s-0.1.0   	1.0
```

**Comment by jawlitkp** [comment]
It is still failing for me. This is the version I am running 
helm version
version.BuildInfo{Version:"v3.1+unreleased", GitCommit:"1911870958098b774973c6fe56bfdf4441f61596", GitTreeState:"clean", GoVersion:"go1.13.10"}

**Comment by nareshnayini** [comment]
This is how I tested this and it worked for me. Only confusing thing for me is it worked without `--replace` flag on `upgrade`. Not sure whether this flag is really required to recover from a failed _only_ release.

**version i am on:**
```
$ helm version
version.BuildInfo{Version:"v3.1+unreleased", GitCommit:"1911870958098b774973c6fe56bfdf4441f61596", GitTreeState:"clean", GoVersion:"go1.13.10"}
```

**First Failed deployment**
```
$ helm upgrade --install helm-test -f deploy/helm/default-values.yaml -f deploy/helm/ci/values.yaml org-helm-template/org-helm
Release "helm-test" does not exist. Installing it now.
Error: ConfigMap "test-helm-$tag" is invalid: metadata.name: Invalid value: "test-helm-$tag": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')

$ helm ls
NAME              NAMESPACE REVISION  UPDATED                               STATUS  CHART               APP VERSION
helm-test  ci        1         2020-04-20 12:53:19.013309 -0400 EDT  failed  org-helm-0.1.1  1.1.0
```

**Successful deployment after revision 1:**
```
$ helm upgrade helm-test -f deploy/helm/default-values.yaml -f deploy/helm/ci/values-test.yaml org-helm-template/org-helm
Release "helm-test" has been upgraded. Happy Helming!
NAME: helm-test
LAST DEPLOYED: Mon Apr 20 12:54:01 2020
NAMESPACE: ci
STATUS: deployed
REVISION: 2

$ helm ls
NAME              NAMESPACE REVISION  UPDATED                             STATUS    CHART               APP VERSION
helm-test  ci        2         2020-04-20 12:54:01.28924 -0400 EDT deployed  org-helm-0.1.1  1.1.0
```

**Comment by yinzara** [comment]
The "--replace" flag was actually removed during the PR review process. It is the default behavior on upgrade to allow upgrading from a failed release.

**Comment by yinzara** [comment]
@hiro5id Your issue is a bit perplexing to me how it could result in the "no deployed releases" error.  If you leave off "--atomic" and "--cleanup-on-fail" (I don't think they're really required anymore in your scenario) does it work?

Was that "helm list" run before you ran the upgrade command? (I'm guessing it's the same before and I after but I wanted to make sure).  Can you also give the output of `helm history`?

**Comment by yinzara** [comment]
@jawlitkp Can you also run both "helm list" and "helm history" and give us that output?

**Comment by nareshnayini** [comment]
```
$ helm list
NAME              NAMESPACE REVISION  UPDATED                             STATUS    CHART               APP VERSION
helm-test  ci        2         2020-04-20 12:54:01.28924 -0400 EDT deployed  org-helm-0.1.1  1.1.0

$ helm history helm-test
REVISION  UPDATED                   STATUS      CHART               APP VERSION DESCRIPTION
1         Mon Apr 20 12:53:19 2020  superseded  org-helm-0.1.1  1.1.0       Release "helm-test" failed: ConfigMap "test-helm-$tag" is invalid: metadata.name: Invalid value: "test-helm-$tag": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
2         Mon Apr 20 12:54:01 2020  deployed    org-helm-0.1.1  1.1.0       Upgrade complete
```

**Comment by yinzara** [comment]
Yeah @nareshnayini  That looks good to me.

**Comment by jawlitkp** [comment]
for me, it is not even keeping good deployment

I always run: helm upgrade --install

REVISION	UPDATED                 	STATUS
1       	Mon Apr 20 16:17:47 2020	supers
2       	Mon Apr 20 16:35:08 2020	deploy
3       	Mon Apr 20 16:42:16 2020	failed
4       	Mon Apr 20 16:43:57 2020	failed
5       	Mon Apr 20 16:45:50 2020	failed



REVISION	UPDATED                 	STATUS
3       	Mon Apr 20 16:42:16 2020	failed
4       	Mon Apr 20 16:43:57 2020	failed
5       	Mon Apr 20 16:45:50 2020	failed
6       	Mon Apr 20 16:47:33 2020	failed
7       	Mon Apr 20 16:49:19 2020	failed

**Comment by bacongobbler** [comment]
@jawlitkp the feature was merged into rc2. Please update and try again. :)

**Comment by yinzara** [comment]
@bacongobbler @jawlitkp The "git commit" from his version statement shows "1911870958098b774973c6fe56bfdf4441f61596" which is the right commit ID to have the "no deployed release" error.   I'm not sure what's going on.

**Comment by jawlitkp** [comment]
Ok, I will just hold for now cuz I broke our ci/cd and had to revert back.

**Comment by virkt25** [comment]
It doesn't seem to work if the release is stuck in a `pending-rollback` state

**Comment by technosophos** [comment]
It will not work for any `pending-*` cases because that status indicates that the state of the release is unknown and not to be trusted. Attempting to construct a patch to fix it could lead to further corruption of the installation.

`pending-*` states occur when an operation makes it mid-way through, but something on the Kubernetes side does not respond or responds in an unanticipated way, and causes a timeout or another premature stopping condition. In those cases, Helm is not equipped to figure out how to reconstruct a correct patch. (The condition is, after all, unknown) So while it is possible that in some cases Helm might be able to recover, it is possible in other cases for Helm to cause unintended (and possibly destructive or monetarily expensive) side effects. So after doing a deep dive into the code, the core maintainers determined that it was not in the community's best interest to support such dangerous behavior, even with a flag.

What we are open to is finding cases where `pending-*` is possibly mis-applied as a state, and fixing that. We currently have no known cases of that, though.

Currently, for releases stuck in "pending" you must either manually restore (which is not something we provide guidance on) or you must delete and recreate.

**Comment by virkt25** [comment]
Thanks @technosophos ! Appreciate the insight. Our release ended up in a weird state I suppose:

```
REVISION        UPDATED                         STATUS                  CHART           APP VERSION     DESCRIPTION                                             

973             Tue Apr 14 16:25:27 2020        failed                  cce-k8s-0.1.0   1.0             Upgrade "cce" failed: timed out waiting for the condition
974             Tue Apr 14 16:30:31 2020        pending-rollback        cce-k8s-0.1.0   1.0             Rollback to 972                                          
975             Tue Apr 14 20:43:40 2020        failed                  cce-k8s-0.1.0   1.0             Upgrade "cce" failed: timed out waiting for the condition
976             Tue Apr 14 20:48:44 2020        pending-rollback        cce-k8s-0.1.0   1.0             Rollback to 972                                          
977             Tue Apr 14 21:36:50 2020        failed                  cce-k8s-0.1.0   1.0             Upgrade "cce" failed: timed out waiting for the condition
978             Tue Apr 14 21:41:54 2020        pending-rollback        cce-k8s-0.1.0   1.0             Rollback to 972                                          
979             Wed Apr 15 02:35:23 2020        failed                  cce-k8s-0.1.0   1.0             Upgrade "cce" failed: timed out waiting for the condition
980             Wed Apr 15 02:40:27 2020        pending-rollback        cce-k8s-0.1.0   1.0             Rollback to 972                                         
981             Thu Apr 16 14:15:42 2020        failed                  cce-k8s-0.1.0   1.0             Upgrade "cce" failed: timed out waiting for the condition
982             Thu Apr 16 14:20:47 2020        pending-rollback        cce-k8s-0.1.0   1.0             Rollback to 972                                          
983             Thu Apr 16 19:51:12 2020        pending-rollback        cce-k8s-0.1.0   1.0             Rollback to 974                                          
984             Thu Apr 16 19:54:13 2020        pending-rollback        cce-k8s-0.1.0   1.0             Rollback to 983                                          
985             Thu Apr 16 19:54:22 2020        pending-rollback        cce-k8s-0.1.0   1.0             Rollback to 981
```

Somehow our build made it so we ended up with a bunch of failed releases and failed rollbacks likely because of a missing secret that because required for a container. Now during this whole time the pods from the last successful deploy continued running because of how we do our rolling-updates (surge by 100% to bring up new containers before old ones are taken down). 

Deleting and recreating a production app is risky and was something we only wanted to do as a last resort and as such we were looking for workarounds. 

I came across this issue: https://github.com/helm/helm/issues/7476 which inspired me to try something and it worked! 

What I did: 
- Edit the last secret storing the helm release to change the label from rollback-pending to deployed. At this point helm status and helm history still showed rollback-pending as the status as it picks it up from the encoded data. I tried to figure out how to decode and update the data but couldn't really figure it out (got it decoded but couldn't re-encode it correctly).
- Using the current canary build I ran `helm upgrade cce ./cce-k8s/ --set image.tag=latest -n namespace`. I had tried this without the above step and it failed but this time much to my surprise it worked!

Documenting this here so it can help others / might help identify a scenario to improve helm further.

**Comment by nareshnayini** [comment]
Hi, is this released as part of `3.2.0` ? I dont see this PR [here](https://github.com/helm/helm/compare/v3.1.0...v3.2.0)

**Comment by hickeyma** [comment]
@nareshnayini This will be in 3.2.1.

**Comment by technosophos** [comment]
I had to re-arrange the 3.2.0 release because of the security issue, and because RC.1 met the release criteria (so there was no need for an RC.2). So I bumped this to 3.2.1.

**Comment by IdanAdar** [comment]
Is it possible to know, programmatically, that the deployment succeeded but that in actuality it deployed/kept the current release because the new release failed and thus this flag kicked the not action?

**Comment by jawlitkp** [comment]
I am planning to roll out a new version before I do just wanted to make sure I am doing the right way.

I am always running below command in our ci-cd pipeline if deploying the first time or upgrading.

 sh "helm upgrade --install ${appName} ${chartPath} --namespace=${namespace} --debug --version ${chartVersion} --set=deployment.image=${image} -f overridesFile.json --history-max=5 --kubeconfig=${kubeConfig} --wait"

Are we expecting above the fix should work for above?

Or 

do I need to change the above command and split based on based if it is installed or upgrades? something like 

if first time install 

sh "helm install ${appName} ${chartPath} --namespace=${namespace} --debug --version ${chartVersion} --set=deployment.image=${image} -f overridesFile.json --history-max=5 --kubeconfig=${kubeConfig} --wait"

And if upgrade then 

sh "helm upgrade ${appName} ${chartPath} --namespace=${namespace} --debug --version ${chartVersion} --set=deployment.image=${image} -f overridesFile.json --history-max=5 --kubeconfig=${kubeConfig} --wait"

**Comment by T1loc** [comment]
Does this MR allow us to use : 

`helm upgrade -i ... ` with a previous uninstalled chart ? 

If I do : 
```
helm list --all --filter frontend                                                                                                                                                                                       
NAME                    NAMESPACE       REVISION        UPDATED                                 STATUS          CHART                   APP VERSION
frontend florian         2               2020-05-27 14:55:56.740062 +0200 CEST   uninstalled     frontend-2.2.0
```

And then I try to upgrae/install it again: 
```
 helm --kube-context dev --version=2.2.0 upgrade -i --reset-values  -f=values.yaml --namespace=florian frontend frontend
Error: UPGRADE FAILED: "frontend" has no deployed releases
```

```
 helm version                                                                                                                                                                                                           
version.BuildInfo{Version:"v3.2.1", GitCommit:"fe51cd1e31e6a202cba7dead9552a6d418ded79a", GitTreeState:"clean", GoVersion:"go1.13.10"}
```

**Comment by bacongobbler** [comment]
No. For that, use `helm install --replace`.

**Pull request merged**
